pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/extensions/manager/SplitManager.java

```diff
@@ -0,0 +1,122 @@
+import org.apache.pulsar.broker.loadbalance.extensions.models.SplitCounter;
+import org.apache.pulsar.broker.loadbalance.extensions.models.SplitDecision;
+
+/**
+ * Split manager.
+ */
+@Slf4j
+public class SplitManager implements StateChangeListener {
+
+    record InFlightSplitRequest(SplitDecision splitDecision, CompletableFuture<Void> future) {
+    }
+
+    private final Map<String, InFlightSplitRequest> inFlightSplitRequests;
+
+    private final SplitCounter counter;
+
+    public SplitManager(SplitCounter splitCounter) {
+        this.inFlightSplitRequests = new ConcurrentHashMap<>();
+        this.counter = splitCounter;
+    }
+
+    private void complete(String serviceUnit, Throwable ex) {
+        inFlightSplitRequests.computeIfPresent(serviceUnit, (__, inFlightSplitRequest) -> {
+            var future = inFlightSplitRequest.future;
+            if (!future.isDone()) {
+                if (ex != null) {
+                    counter.update(Failure, Unknown);
+                    future.completeExceptionally(ex);
+                    log.error("Failed the bundle split event: {}", serviceUnit, ex);
+                } else {
+                    counter.update(inFlightSplitRequest.splitDecision);
+                    future.complete(null);
+                    log.info("Completed the bundle split event: {}", serviceUnit);
+                }
+            }
+            return null;
+        });
+    }
+
+    public CompletableFuture<Void> waitAsync(CompletableFuture<Void> eventPubFuture,
+                                             String bundle,
+                                             SplitDecision decision,
+                                             long timeout,
+                                             TimeUnit timeoutUnit) {
+        return eventPubFuture
+                .thenCompose(__ -> inFlightSplitRequests.computeIfAbsent(bundle, ignore -> {
+                    log.info("Published the bundle split event for bundle:{}. "
+                                    + "Waiting the split event to complete. Timeout: {} {}",
+                            bundle, timeout, timeoutUnit);
+                    CompletableFuture<Void> future = new CompletableFuture<>();
+                    future.orTimeout(timeout, timeoutUnit).whenComplete((v, ex) -> {
+                        if (ex != null) {
+                            inFlightSplitRequests.remove(bundle);
+                            log.warn("Timed out while waiting for the bundle split event: {}", bundle, ex);
+                        }
+                    });
+                    return new InFlightSplitRequest(decision, future);
+                }).future)
+                .exceptionally(e -> {
+                    log.error("Failed to publish the bundle split event for bundle:{}. Skipping wait.", bundle);
+                    counter.update(Failure, Unknown);
+                    return null;
```

Comment:
Using exceptionally will cause the returned future to lose the exception info. We should use `whenComplete` here.

Reply:
Updated.